File: Implementations/Swift/Sources/AstrolabeProtocol/Inspection/Hierarchy/RuntimeNodeRelation.swift
//
//  RuntimeNodeRelation.swift
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/8/4.
//

/// Directed platform-neutral relation between two captured Runtime nodes.
public struct RuntimeNodeRelation: Codable, Equatable, Sendable {
    /// Open namespaced relation type owned by its producer.
    public let type: RuntimeNamespacedIdentifier

    /// Source node identifier for this directed relation.
    public let sourceNodeID: RuntimeOpaqueIdentifier

    /// Target node identifier for this directed relation.
    public let targetNodeID: RuntimeOpaqueIdentifier

    /// Namespaced producer-specific relation facts.
    public let extensions: RuntimeExtensionMap

    public init(
        type: RuntimeNamespacedIdentifier,
        sourceNodeID: RuntimeOpaqueIdentifier,
        targetNodeID: RuntimeOpaqueIdentifier,
        extensions: RuntimeExtensionMap
    ) {
        self.type = type
        self.sourceNodeID = sourceNodeID
        self.targetNodeID = targetNodeID
        self.extensions = extensions
    }
}

File: Implementations/Kotlin/src/main/kotlin/dev/astrolabe/protocol/inspection/RuntimeNodeRelation.kt
//
//  RuntimeNodeRelation.kt
//  astrolabe-protocol
//
//  Created by 轩辕十四 on 2026/8/4.
//

package dev.astrolabe.protocol

import kotlinx.serialization.Serializable

/** Directed platform-neutral relation between two captured Runtime nodes. */
@Serializable
public data class RuntimeNodeRelation(
    /** Open namespaced relation type owned by its producer. */
    public val type: RuntimeNamespacedIdentifier,
    /** Source node identifier for this directed relation. */
    public val sourceNodeID: RuntimeOpaqueIdentifier,
    /** Target node identifier for this directed relation. */
    public val targetNodeID: RuntimeOpaqueIdentifier,
    /** Namespaced producer-specific relation facts. */
    public val extensions: RuntimeExtensionMap
)
